### PR TITLE
CNTRLPLANE-3624: e2e v2 test for control-plane-pki-operator TLS configuration

### PR DESCRIPTION
**Pull request by kaleemsiddiqu**
[CNTRLPLANE-3624](https://redhat.atlassian.net/browse/CNTRLPLANE-3624): Add end-to-end v2 test to verify control-plane-pki-operator TLS configuration behavior.

## What this PR does / why we need it:
e2e test for https://github.com/openshift/hypershift/pull/8768

## Summary by CodeRabbit

* **Tests**
  * Added an end-to-end test suite that validates control-plane-pki-operator TLS behavior when the HostedCluster API server `TLSSecurityProfile` changes, including checks for required `minTLSVersion` values and connectivity success or failure for TLS 1.2 vs TLS 1.3 across default/intermediate and Modern scenarios.

### CODE REVIEW

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+			Expect(configYAML).NotTo(BeEmpty(), "config.yaml should not be empty")
+			Expect(configYAML).To(ContainSubstring("TLS"), "config.yaml should contain TLS configuration")
+		})
+
+		It("should have minTLSVersion set to VersionTLS12 with default/intermediate profile", func() {
+			// Check HostedCluster TLS profile configuration
+			hostedCluster := tc.GetHostedCluster()
+
+			// Check if profile is default or intermediate (both support TLS 1.2)
+			if hostedCluster.Spec.Configuration != nil &&
```
Comment (csrwng) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go]: **Fragile guard: "not Modern" != "supports TLS 1.2"**

This skip guard uses `!= configv1.TLSProfileModernType` as a proxy for "should have minTLSVersion TLS 1.2", but if a future profile type is introduced (e.g., Custom with TLS 1.3-only ciphers), this test would run and incorrectly assert `VersionTLS12`. Consider guarding on the positive case instead — skip unless the profile is nil (default) or explicitly `IntermediateType`.

The same pattern applies at lines 125-129.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+		})
+
+		It("should accept both TLS 1.2 and TLS 1.3 connections with intermediate profile", func() {
+			// Check HostedCluster TLS profile configuration
+			hostedCluster := tc.GetHostedCluster()
+
+			if hostedCluster.Spec.Configuration != nil &&
+				hostedCluster.Spec.Configuration.APIServer != nil &&
+				hostedCluster.Spec.Configuration.APIServer.TLSSecurityProfile != nil &&
+				hostedCluster.Spec.Configuration.APIServer.TLSSecurityProfile.Type == configv1.TLSProfileModernType {
```
Comment (csrwng) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go]: **Nit: extract the TLS profile type check into a helper**

This same multi-line nil-check-then-compare pattern appears in at least 4 tests (lines 95-99, 125-129, 239-242, 319-322). Consider extracting it into a small helper like `hostedClusterHasTLSProfileType(hc, profileType) bool` to reduce repetition and make the skip guards more readable.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+				Skip("HostedCluster has Modern TLS profile - skipping intermediate profile connectivity test")
+			}
+
+			// Get control-plane-pki-operator pod from management cluster
+			mgmtClient := tc.MgmtClient
+			mgmtRestConfig, err := e2eutil.GetConfig()
```
Comment (csrwng) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go]: **Nit: extract mgmt REST config and kube client setup**

`e2eutil.GetConfig()` + `kubernetes.NewForConfig()` is repeated in 3 tests (here, line 247, line 389). Consider setting these up once in `BeforeAll` alongside `tc` and storing them as shared variables, or extracting into a helper.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+
+			pkiPod := &pkiPodList.Items[0]
+			Expect(pkiPod.Status.Phase).To(Equal(corev1.PodRunning),
+				"control-plane-pki-operator pod should be running")
+
+			metricsPort := "8443"
+			GinkgoWriter.Printf("Testing TLS connections to PKI operator pod %s on port %s\n", pkiPod.Name, metricsPort)
+
+			// Test TLS 1.2 connection using openssl s_client from within the PKI operator pod
+			tls12Result, _ := v2util.RunCommandInPod(tc.Context, mgmtKubeClient, mgmtRestConfig,
```
Comment (csrwng) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go]: **Bug risk: `RunCommandInPod` error is silently discarded**

The error return is ignored here and in every other `RunCommandInPod` call in this file (lines 157, 165, 280, 288, 422, 430). While `|| true` prevents a non-zero exit from the shell command itself, an infrastructure-level failure (e.g., pod exec API error, container not found) would also be silently swallowed, and the test would assert against an empty or garbage string. Consider checking the error:
```go
tls12Result, err := v2util.RunCommandInPod(...)
Expect(err).NotTo(HaveOccurred(), "failed to exec into PKI operator pod")
```

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+
+func RegisterControlPlanePKIOperatorTests(getTestCtx internal.TestContextGetter) {
+	VerifyPKIOperatorTLSConfigTest(getTestCtx)
+}
+
+// VerifyPKIOperatorTLSConfigTest validates that when TLS security profile changes are applied,
+// the control-plane-pki-operator config reflects the correct minTLSVersion and that
+// the API server enforces those TLS versions correctly.
+func VerifyPKIOperatorTLSConfigTest(getTestCtx internal.TestContextGetter) {
+	When("PKI operator TLS configuration is modified", Ordered, func() {
```
Comment (csrwng) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go]: **Add `Label("Lifecycle")`: this container includes tests that mutate the HostedCluster**

Per the v2 framework guidelines (Rule 13), only lifecycle tests may mutate cluster state. The `It` blocks at lines 173 and 307 update the HostedCluster's TLS security profile. Adding `Label("Lifecycle")` here ensures CI can filter these appropriately (longer timeouts, avoid parallel execution with other mutation tests):
```suggestion
	When("PKI operator TLS configuration is modified", Ordered, Label("Lifecycle"), func() {
```

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+			// Test TLS 1.3 connection using openssl s_client
+			tls13Result, _ := v2util.RunCommandInPod(tc.Context, mgmtKubeClient, mgmtRestConfig,
+				tc.ControlPlaneNamespace, pkiPod.Name, "control-plane-pki-operator",
+				"sh", "-c",
+				fmt.Sprintf("timeout 2 openssl s_client -connect localhost:%s -tls1_3 2>&1 || true", metricsPort))
+			Expect(strings.ToLower(tls13Result)).To(ContainSubstring("tlsv1.3"),
+				"should confirm TLS 1.3 was used for PKI operator port %s", metricsPort)
+		})
+
+		It("should update to minTLSVersion VersionTLS13 when TLS profile changed to Modern", func() {
```
Comment (csrwng) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go:186]: This test mutates the HostedCluster (sets TLS profile to Modern). The containing `When` block should carry `Label("Lifecycle")` per framework Rule 13.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+					ContainSubstring("handshake failure"),
+					ContainSubstring("protocol version"),
+					ContainSubstring("error"),
+					ContainSubstring("wrong version number"),
+					Not(ContainSubstring("protocol  : tlsv1.2")),
+				),
+				"TLS 1.2 connection should be rejected with modern profile")
+		})
+
+		It("should downgrade to minTLSVersion VersionTLS12 when Modern TLS profile is removed", func() {
```
Comment (csrwng) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go:365]: This test mutates the HostedCluster (removes TLS profile to downgrade to Intermediate). The containing `When` block should carry `Label("Lifecycle")` per framework Rule 13.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+				if hostedCluster.Spec.Configuration == nil {
+					hostedCluster.Spec.Configuration = &hyperv1.ClusterConfiguration{}
+				}
+				if hostedCluster.Spec.Configuration.APIServer == nil {
+					hostedCluster.Spec.Configuration.APIServer = &configv1.APIServerSpec{}
+				}
+
+				// Update to Modern TLS profile in the HostedCluster CR
+				hostedCluster.Spec.Configuration.APIServer.TLSSecurityProfile = &configv1.TLSSecurityProfile{
+					Type:   configv1.TLSProfileModernType,
```
Comment (csrwng) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go:219]: **Bug: conflict-retry pattern silently succeeds without applying the update**

`g.Expect(err).To(HaveOccurred(), ...)` on a non-nil `err` always passes, so when a conflict occurs this `Eventually` iteration returns *successfully* — `Eventually` considers the body satisfied and stops retrying. The update was never applied, but the test proceeds as if it was.

To actually retry on conflict, you need to fail the iteration explicitly:
```suggestion
				err = mgmtClient.Update(tc.Context, hostedCluster)
				g.Expect(apierrors.IsConflict(err)).To(BeFalse(), "conflict on update, retrying")
				g.Expect(err).NotTo(HaveOccurred(), "failed to update HostedCluster TLS profile to Modern")
```

The same bug appears at line 335-337 in the downgrade test.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+					crclient.InNamespace(tc.ControlPlaneNamespace),
+					crclient.MatchingLabels{"app": "control-plane-pki-operator"},
+				)).To(Succeed(), "failed to list control-plane-pki-operator pods")
+
+				g.Expect(pkiPodList.Items).NotTo(BeEmpty(),
+					"expected at least one control-plane-pki-operator pod")
+
+				pkiPod = &pkiPodList.Items[0]
+				g.Expect(pkiPod.Status.Phase).To(Equal(corev1.PodRunning),
+					"control-plane-pki-operator pod should be running")
```
Comment (csrwng) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go:286]: **Vacuous pass: PodReady condition may not exist**

Per framework Rule 16, condition-search loops need a `found` guard. If the pod has no `PodReady` condition in its status (e.g., during init), no assertion fires and the `Eventually` iteration succeeds without actually verifying readiness. Same issue at lines 410-414.

```suggestion
				readyFound := false
				for _, cond := range pkiPod.Status.Conditions {
					if cond.Type == corev1.PodReady {
						readyFound = true
						g.Expect(cond.Status).To(Equal(corev1.ConditionTrue),
							"control-plane-pki-operator pod should be ready")
					}
				}
				g.Expect(readyFound).To(BeTrue(), "PodReady condition not found on pod %s", pkiPod.Name)
```

**Review by csrwng — COMMENTED** [review]
Review of PR #8793 — e2e v2 test for control-plane-pki-operator TLS configuration.

Overall the test structure follows the v2 framework conventions well. Found one likely bug in the conflict-handling pattern (appears twice), a cleanup gap in `AfterAll`, and several robustness/maintainability improvements.

11 comments total: 1 bug, 1 cleanup gap, 9 suggestions.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+				"should confirm TLS 1.3 was used for PKI operator with Modern profile")
+
+			// Test TLS 1.2 connection should fail using openssl
+			tls12Result, _ := v2util.RunCommandInPod(tc.Context, mgmtKubeClient, mgmtRestConfig,
+				tc.ControlPlaneNamespace, pkiPod.Name, "control-plane-pki-operator",
+				"sh", "-c",
+				fmt.Sprintf("timeout 2 openssl s_client -connect localhost:%s -tls1_2 2>&1 || true", metricsPort))
+
+			// TLS 1.2 should be rejected - expect error or alert in output, or no TLSv1.2 in successful output
+			lowerResult := strings.ToLower(tls12Result)
```
Comment (csrwng) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go:360]: **Overly broad assertion: `ContainSubstring("error")` matches successful connections too**

The `Or()` matcher has several issues:
1. `ContainSubstring("error")` is too broad — even a successful openssl handshake outputs debug lines containing "error" (e.g., `Verify return code: ...` preceded by error-related lines, `error writing to BIO`). This arm would match on a *successful* TLS 1.2 connection, making the test always pass.
2. `ContainSubstring("alert")` is also common in successful TLS output.
3. `Not(ContainSubstring("protocol  : tlsv1.2"))` with the double space is openssl-version-dependent formatting.

Consider asserting the negative instead — that the connection was NOT successfully negotiated at TLS 1.2:
```go
Expect(lowerResult).NotTo(
    And(
        ContainSubstring("new,"),
        ContainSubstring("tlsv1.2"),
    ),
    "TLS 1.2 connection should be rejected with modern profile, got: %s", tls12Result)
```
Or check for the specific openssl error that TLS 1.3-only servers return: `ContainSubstring("no protocols available")` or `ContainSubstring("unsupported protocol")`.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+				tc.ControlPlaneNamespace, pkiPod.Name, "control-plane-pki-operator",
+				"sh", "-c",
+				fmt.Sprintf("timeout 2 openssl s_client -connect localhost:%s -tls1_3 2>&1 || true", metricsPort))
+			Expect(strings.ToLower(tls13Result)).To(ContainSubstring("tlsv1.3"),
+				"should confirm TLS 1.3 was accepted for PKI operator with Intermediate profile")
+		})
+
+		AfterAll(func() {
+			// Restore original TLS profile if it was captured
+			if originalTLSProfile != nil && tc != nil {
```
Comment (csrwng) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go]: **Cleanup gap: `AfterAll` doesn't restore when the original TLS profile was nil**

If the HostedCluster starts with no TLS profile (the common default — `originalTLSProfile` is nil), this cleanup block is entirely skipped. If the test fails after setting Modern (line 195) but before the downgrade test runs (line 307), `AfterAll` won't restore the HC — it stays mutated with Modern. With `Ordered`, a failure in any earlier `It` causes later ones to be skipped, so this is a realistic scenario.

Consider always running the cleanup and handling the nil case:
```suggestion
		AfterAll(func() {
			if tc == nil {
				return
			}
			GinkgoWriter.Printf("Restoring original TLS security profile\n")

			err := wait.PollUntilContextTimeout(tc.Context, 5*time.Second, 1*time.Minute, true, func(ctx context.Context) (bool, error) {
				hostedCluster := &hyperv1.HostedCluster{}
				err := tc.MgmtClient.Get(ctx, crclient.ObjectKey{
					Namespace: tc.ClusterNamespace,
					Name:      tc.ClusterName,
				}, hostedCluster)
				if err != nil {
					return false, err
				}

				if hostedCluster.Spec.Configuration == nil {
					hostedCluster.Spec.Configuration = &hyperv1.ClusterConfiguration{}
				}
				if hostedCluster.Spec.Configuration.APIServer == nil {
					hostedCluster.Spec.Configuration.APIServer = &configv1.APIServerSpec{}
				}
				hostedCluster.Spec.Configuration.APIServer.TLSSecurityProfile = originalTLSProfile
```
This way, if `originalTLSProfile` is nil, it sets the profile back to nil (restoring the default).

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+		hc.Spec.Configuration.APIServer.TLSSecurityProfile.Type == profileType
+}
+
+func RegisterControlPlanePKIOperatorTests(getTestCtx internal.TestContextGetter) {
+	VerifyPKIOperatorTLSConfigTest(getTestCtx)
+}
+
+// VerifyPKIOperatorTLSConfigTest validates that when TLS security profile changes are applied,
+// the control-plane-pki-operator config reflects the correct minTLSVersion and that
+// the API server enforces those TLS versions correctly.
```
Comment (ricardomaraschini) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go]: What API server are we referring to here ?

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 13]
```diff
@@ -0,0 +1,499 @@
...
+			// Check in management cluster's control plane namespace, not hosted cluster
+			mgmtClient := tc.MgmtClient
+			cm := &corev1.ConfigMap{}
+			err := mgmtClient.Get(tc.Context, crclient.ObjectKey{
+				Namespace: tc.ControlPlaneNamespace,
+				Name:      pkiOperatorConfigMapName,
+			}, cm)
+
+			if apierrors.IsNotFound(err) {
+				Skip(fmt.Sprintf("ConfigMap %s/%s not found - this test requires the PKI operator config to be present",
```
Comment (ricardomaraschini) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go]: Why isn't this a failure instead ?

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 14]
```diff
@@ -0,0 +1,499 @@
...
+
+			// Only run for nil (default) or explicit Intermediate profile
+			hasProfile := hostedCluster.Spec.Configuration != nil &&
+				hostedCluster.Spec.Configuration.APIServer != nil &&
+				hostedCluster.Spec.Configuration.APIServer.TLSSecurityProfile != nil
+
+			isDefaultOrIntermediate := !hasProfile || hostedClusterHasTLSProfileType(hostedCluster, configv1.TLSProfileIntermediateType)
+
+			if !isDefaultOrIntermediate {
+				Skip("HostedCluster does not have default or Intermediate TLS profile - skipping TLS 1.2 test")
```
Comment (ricardomaraschini) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go:118]: Why we keep skipping tests ? Shouldn't we make sure the transitions work ?

**Review by ricardomaraschini — COMMENTED** [review]
Have you considered making this part of the existing test on the `origin` repository ?

**Comment by kaleemsiddiqu** [comment]
> Have you considered making this part of the existing test on the `origin` repository ?

Those tests are very broad and covers things in general involving all related operators on both standard cluster and hypershift, this is specific to hypershift only thats why i thought to keep the test for this here only.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go (hunk 14 above)
Comment (kaleemsiddiqu) [reply test/e2e/v2/tests/control_plane_pki_operator_test.go:118]: This is just an additional check here before the transition (upgrade/downgrade) in configmap itself before upgrade transition

File: test/e2e/v2/tests/control_plane_pki_operator_test.go (hunk 13 above)
Comment (kaleemsiddiqu) [reply test/e2e/v2/tests/control_plane_pki_operator_test.go]: Added this as failure.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go (hunk 12 above)
Comment (kaleemsiddiqu) [reply test/e2e/v2/tests/control_plane_pki_operator_test.go]: Updated text with better description.

**Review by jparrill — COMMENTED** [review]
Dropped some comments. Thanks!

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 15]
```diff
@@ -0,0 +1,556 @@
...
+}
+
+func RegisterControlPlanePKIOperatorTests(getTestCtx internal.TestContextGetter) {
+	VerifyPKIOperatorTLSConfigTest(getTestCtx)
+}
+
+// VerifyPKIOperatorTLSConfigTest validates that when TLS security profile changes are applied
+// to the HostedCluster, the control-plane-pki-operator config reflects the correct minTLSVersion
+// and that the control-plane-pki-operator's HTTPS endpoint enforces those TLS versions correctly.
+func VerifyPKIOperatorTLSConfigTest(getTestCtx internal.TestContextGetter) {
```
Comment (jparrill) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go:62]: Needs `Serial` here too. This test mutates the HC TLS profile and triggers pod restarts — without `Serial`, parallel tests checking pod health or certs will flake. See `backup_restore_test.go` for the pattern.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 16]
```diff
@@ -0,0 +1,556 @@
...
+			Expect(cm.Data).NotTo(BeNil(), "ConfigMap data should not be nil")
+			Expect(cm.Data).To(HaveKey("config.yaml"), "ConfigMap should have config.yaml key")
+
+			// Verify config contains TLS-related fields
+			configYAML := cm.Data["config.yaml"]
+			Expect(configYAML).NotTo(BeEmpty(), "config.yaml should not be empty")
+			Expect(configYAML).To(ContainSubstring("TLS"), "config.yaml should contain TLS configuration")
+		})
+
+		It("should have minTLSVersion set to VersionTLS12 with default/intermediate profile", func() {
```
Comment (jparrill) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go:106]: This assertion is meaningless — "TLS" appears everywhere in field names and cipher strings. The subsequent tests already validate the real config. Remove this line.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 17]
```diff
@@ -0,0 +1,556 @@
...
+
+			isDefaultOrIntermediate := !hasProfile || hostedClusterHasTLSProfileType(hostedCluster, configv1.TLSProfileIntermediateType)
+
+			if !isDefaultOrIntermediate {
+				Skip("HostedCluster does not have default or Intermediate TLS profile - skipping TLS 1.2 connectivity test")
+			}
+
+			// Get control-plane-pki-operator pod from management cluster
+			mgmtClient := tc.MgmtClient
+			pkiPodList := &corev1.PodList{}
```
Comment (jparrill) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go:152]: Declared 3 times as a local variable. Hoist to a constant next to `pkiOperatorConfigMapName`.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 18]
```diff
@@ -0,0 +1,556 @@
...
+
+			pkiPod := &pkiPodList.Items[0]
+			Expect(pkiPod.Status.Phase).To(Equal(corev1.PodRunning),
+				"control-plane-pki-operator pod should be running")
+
+			metricsPort := "8443"
+			GinkgoWriter.Printf("Testing TLS connections to PKI operator pod %s on port %s\n", pkiPod.Name, metricsPort)
+
+			// Test TLS 1.2 connection using openssl s_client from within the PKI operator pod
+			tls12Result, err := v2util.RunCommandInPod(tc.Context, mgmtKubeClient, mgmtRestConfig,
```
Comment (jparrill) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go:168]: `g.Expect(apierrors.IsConflict(err)).To(BeFalse())` works by accident but reads wrong. Use:
```go
if apierrors.IsConflict(err) {
    return
}
g.Expect(err).NotTo(HaveOccurred())
```
Same at lines 287-288.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 19]
```diff
@@ -0,0 +1,556 @@
...
+			if !hostedClusterHasTLSProfileType(hostedCluster, configv1.TLSProfileModernType) {
+				Fail("HostedCluster does not have Modern TLS profile - previous ordered test should have set it")
+			}
+
+			// Wait for PKI operator pod to restart and pick up the new TLS config
+			Eventually(func(g Gomega) {
+				pkiPodList := &corev1.PodList{}
+				g.Expect(mgmtClient.List(tc.Context, pkiPodList,
+					crclient.InNamespace(tc.ControlPlaneNamespace),
+					crclient.MatchingLabels{"app": "control-plane-pki-operator"},
```
Comment (jparrill) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go]: You wait for a pod to be Ready, but never verify it is a *new* pod. The old pod with old config can still be Running+Ready briefly. TLS 1.3 passes against both old and new config, so you get a false positive. Capture pod UID before mutation, wait for a different UID after.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 20]
```diff
@@ -0,0 +1,556 @@
...
+
+				// Check Ready condition
+				readyFound := false
+				for _, cond := range pkiPod.Status.Conditions {
+					if cond.Type == corev1.PodReady {
+						readyFound = true
+						g.Expect(cond.Status).To(Equal(corev1.ConditionTrue),
+							"control-plane-pki-operator pod should be ready")
+					}
+				}
```
Comment (jparrill) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go:296]: `timeout 2` is pretty aggressive for CI. The pod exec goes through SPDY streaming via the API server — in a loaded cluster, 2 seconds might not be enough. Bump to `timeout 5` across all the openssl commands.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go [hunk 21]
```diff
@@ -0,0 +1,556 @@
...
+			GinkgoWriter.Printf("Restoring original TLS security profile\n")
+
+			// Use Eventually with retry on conflict
+			err := wait.PollUntilContextTimeout(tc.Context, 5*time.Second, 1*time.Minute, true, func(ctx context.Context) (bool, error) {
+				hostedCluster := &hyperv1.HostedCluster{}
+				err := tc.MgmtClient.Get(ctx, crclient.ObjectKey{
+					Namespace: tc.ClusterNamespace,
+					Name:      tc.ClusterName,
+				}, hostedCluster)
+				if err != nil {
```
Comment (jparrill) [inline test/e2e/v2/tests/control_plane_pki_operator_test.go]: Two things:

1. **`wait.PollUntilContextTimeout`**: No other v2 test uses this — the framework convention is `Eventually`. You already use `Eventually` with conflict retry earlier in the same file. Switching also lets you drop the `"context"` and `wait` imports.

2. **Silent failure**: The warning log swallows the restore error. If this fails, the cluster is left with whatever TLS profile was last set, breaking subsequent tests. Use `Expect(err).NotTo(HaveOccurred())` to fail loudly.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go (hunk 15 above)
Comment (kaleemsiddiqu) [reply test/e2e/v2/tests/control_plane_pki_operator_test.go:62]: Added `serial` here

File: test/e2e/v2/tests/control_plane_pki_operator_test.go (hunk 16 above)
Comment (kaleemsiddiqu) [reply test/e2e/v2/tests/control_plane_pki_operator_test.go:106]: Removed that check

File: test/e2e/v2/tests/control_plane_pki_operator_test.go (hunk 17 above)
Comment (kaleemsiddiqu) [reply test/e2e/v2/tests/control_plane_pki_operator_test.go:152]: Moved as constant and replaced wherever it was used.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go (hunk 18 above)
Comment (kaleemsiddiqu) [reply test/e2e/v2/tests/control_plane_pki_operator_test.go:168]: Updated with suggested code

File: test/e2e/v2/tests/control_plane_pki_operator_test.go (hunk 19 above)
Comment (kaleemsiddiqu) [reply test/e2e/v2/tests/control_plane_pki_operator_test.go]: added code to compare pods UID before and after.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go (hunk 20 above)
Comment (kaleemsiddiqu) [reply test/e2e/v2/tests/control_plane_pki_operator_test.go:296]: timeout of 5 secods added.

File: test/e2e/v2/tests/control_plane_pki_operator_test.go (hunk 21 above)
Comment (kaleemsiddiqu) [reply test/e2e/v2/tests/control_plane_pki_operator_test.go]: 1. Replaced with Eventually
2. Used Expect(err).*

**Comment by kaleemsiddiqu** [comment]
/verified by @kaleemsiddiqu 
Local execution is successful.
